### PR TITLE
Add mac0 support

### PR DESCRIPTION
**Pull request by adam2809**
Adds COSE_Mac0 support and tests for it.

### CODE REVIEW

File: Makefile.test
Comment (laurencelundblade) [inline Makefile.test:81]: t_cose_parameter.h is now a public header

File: Makefile.ossl
Comment (laurencelundblade) [inline Makefile.ossl:136]: t_cose_parameter.h is no public in the inc/t_cose directory

File: src/t_cose_mac_verify.c
Comment (laurencelundblade) [inline src/t_cose_mac_verify.c:126]: I think the whole short-circuit thing can be eliminated for MAC. It was there for signing because the pubic key crypto is harder to integrate and test with and because the keys are harder to generate and such. This is much less true of HMAC, so I think it can be removed.

File: src/t_cose_mac_verify.c
Comment (laurencelundblade) [inline src/t_cose_mac_verify.c:221]: There's a whole new API for decoding header parameters in t_cose_parameters.h. This should probably switch to that to be able to handle custom parameters and so that COSE_Recipients can be part of MAC at some point.

**Review by laurencelundblade — CHANGES_REQUESTED** [review]
So I'd like to merge this sooner rather than later. To merge it I want to be sure the file names are right and the testing are passing. Other details can be addressed in follow-on PRs.

I'd like to rename to t_cose_mac_compute and t_cose_mac_validate or t_cose_mac_check.  This is to be clearly separate from signing.

The other thing fix before merging is the referent to t_cose_parameters.h. You can still use the old parameter encoding and decoding, as it is still there, but note that t_cose_parameters.h is now public. 

I've made a bunch of other comments in the review here. It's nice to address some of them before merging, but they don't have to be all addressed. 

This all generally looks very good and thoroughly done. Thank you!

**Comment by tamasban** [comment]
@laurencelundblade Thanks for the review. @adam2809 will update the PR when he returned from his holiday at the beginning of August.

File: Makefile.ossl
Comment (adam2809) [reply Makefile.ossl:57]: Yes some functions in OpenSSL are not implemented in the crypto adapter from line 641 onwards as well as some helper functions in tests.

File: inc/t_cose/t_cose_common.h
Comment (adam2809) [reply inc/t_cose/t_cose_common.h:17]: It is because of inc/t_cose/t_cose_common.h:680:36

Comment (laurencelundblade) [reply inc/t_cose/t_cose_common.h:17]: qcbor_decode_error_to_t_cose_error() should go in t_cose_util.h. This function is not a part of the t_cose public API.

File: src/t_cose_mac_verify.c
Comment (adam2809) [reply src/t_cose_mac_verify.c:221]: I can see that the sign1 verify part of the code is adjested to the now API bu not fully I mean the context structure is not updated and some other things. I got mac verify part to the same level and I am leaving it for now since I expect things might sill change and I could really use a working example.

Comment (laurencelundblade) [reply src/t_cose_mac_verify.c:221]: I saw a comment from Adam in email, but can't find it here in GitHub. 

I want to respond and say I think it is OK to merge COSE_Mac even if he parameter stuff is the old parameter stuff. If you fix the merge conflicts and get the tests running we can merge Mac.

File: inc/t_cose/t_cose_mac_verify.h
Comment (adam2809) [reply inc/t_cose/t_cose_mac_verify.h:99]: I had way more debug related issues than I expected which is the reason that resolving this stuff took so long I managed to resolve everything but this since this part will require changing the mac ctx struct which I exepect will cascade into lots of more changes. For now feel free to review what I did for now.

File: test/t_cose_sign_verify_mac_test.c
Comment (laurencelundblade) [inline test/t_cose_sign_verify_mac_test.c:19]: This should be the test functions for MAC for all crypto libraries, so there shouldn't be a PSA dependency here.

File: test/t_cose_sign_verify_mac_test.c
Comment (laurencelundblade) [inline test/t_cose_sign_verify_mac_test.c:28]: cose algs should always be int32_t in t_cose

**Review by laurencelundblade — CHANGES_REQUESTED** [review]
I've made some requests for renaming files and functions here. I think we want to settle that before the merge.

Also the merge conflicts have to be fixed.

I haven't tried to compile or run because of the merge conflicts, but will when they are fixed. Want to make sure all the tests pass before merging.

Note that I used "tdv/b.sh" as the full test before merges. It is too complicated and long-running for CI, so I just run it manually.

File: src/t_cose_util.h
Comment (adam2809) [reply src/t_cose_util.h:64]: I think it might be useful later on for detached payload so I think it should be left for now

**Comment by adam2809** [comment]
Rebased my branch on top of dev to remove conlicts @laurencelundblade feel free to merge it if you feel it is ok.

File: src/t_cose_mac_validate.c
Comment (laurencelundblade) [inline src/t_cose_mac_validate.c:129]: Should be {4, params_arr}.  4 is the size of the array that can be used.  4 should be T_COSE_NUM_VERIFY_DECODE_HEADERS
params must be part of t_cose_mac_validate_ctx so you can return it to the caller.

File: src/t_cose_mac_validate.c
Comment (laurencelundblade) [inline src/t_cose_mac_validate.c:238]: returning address of a stack variable. This will go badly...

**Review by laurencelundblade — CHANGES_REQUESTED** [review]
Please compile with /usr/local/bin/gcc-11 -Wall -Wextra -Wpedantic -Wshadow -Wconversion -Wcast-qual -std=c99 -xc -Wstrict-prototypes.  It's not in the distributed makefiles to avoid trouble for people that download, but I do build with them before releases. There are some warnings to fix.

might have further requests...

File: Makefile.psa
Comment (adam2809) [reply Makefile.psa:107]: Well I don't really know what to do here it doesn't work at all with spaces on my ubuntu machine.

**Comment by laurencelundblade** [comment]
Hello,

Maybe you are not done yet, but I did check out the latest.  There are still lots of warnings when compiled with -Wall -Wextra -Wpedantic -Wshadow -Wconversion -Wcast-qual -std=c99 -xc -Wstrict-prototypes

Also the resolution on the returning of the parameters is to move the parameter storage into the mac verify context so it can be returned since it needs to be returned. This is what sign verify does.

On the spaces/tabs in the makefile, what version of Make are you using? I have GNU Make 3.81

**Comment by adam2809** [comment]
> Hello,
> 
> Maybe you are not done yet, but I did check out the latest. There are still lots of warnings when compiled with -Wall -Wextra -Wpedantic -Wshadow -Wconversion -Wcast-qual -std=c99 -xc -Wstrict-prototypes
> 
> Also the resolution on the returning of the parameters is to move the parameter storage into the mac verify context so it can be returned since it needs to be returned. This is what sign verify does.
> 
> On the spaces/tabs in the makefile, what version of Make are you using? I have GNU Make 3.81
> 
> Still waiting to get all tests to pass so the build doesn't break when this merges...

I don't really understand what is the point of returning the parameters explicitly if the struct already has them. If the user of the API wants to access them why not just get the pointer from the struct? Is this just for backwards compatibility?

**Comment by laurencelundblade** [comment]
The reason for access through a function is for the clear separation of public APIs that should never change and private data structures and such that can change.  

In a lot of cases the accessor function is inline so there ends up being no code generated.

t_cose is being a bit object oriented even though it is in C. If this were C++ we could really make the data structure private.

Access through the API also makes it clear when the data is valid.

**Comment by laurencelundblade** [comment]
Glad to see the CI tests passing.

Right now makefiles are in some trouble with necessary tabs being replaced by spaces so I can't check the latest stuff out.

**Comment by adam2809** [comment]
> Glad to see the CI tests passing.
> 
> Right now makefiles are in some trouble with necessary tabs being replaced by spaces so I can't check the latest stuff out.

It should work now because of the RECIPEPREFIX variable on top it should set the characters which are considered indentation by make. If it doesn't work try using a newer make version.

**Comment by laurencelundblade** [comment]
I'd really like the makefile changes to go back to be the absolute minimum needed to add Mac0. In particular have no changes to tabs and spaces. This will really help get this merged.

**Pull request merged**
